Component controller

drpicox opened the pull request:
See PR: https://github.com/angular/angular.js/pull/13732
See Issue: #13683 

Now `$componentController` looks for the directive that represents a component which is the one that satisfies:
- it has at least restrict:'E' (other are valid if they contain 'E')
- it has controller
- it has controllerAs.

Tests are added to evaluate under multiple scenarios commented in the original thread.

PS: I created the new PR to solve conflicts with minimum "_dirtiness_".

drpicox committed:
feat(ngMock.$componentController): use $injector instead of adding new code to angular.min.js #13683

drpicox committed:
feat(ngMock.$componentController): handle multiple directives with the same name #13683

petebacondarwin commented on src/ngMock/angular-mocks.js:
```diff
@@ -2183,10 +2183,25 @@ angular.mock.$ControllerDecorator = ['$delegate', function($delegate) {
+        // look for those directives that are components
+        var candidateDirectives = directives.filter(function(directiveInfo) {
+          // components have controller, controllerAs and restrict:'E' compatible
+          return directiveInfo.controller && directiveInfo.controllerAs && directiveInfo.restrict.indexOf('E') >= 0;
```
I think we can just check `=== 'E'`

petebacondarwin commented on src/ngMock/angular-mocks.js:
```diff
@@ -2183,10 +2183,25 @@ angular.mock.$ControllerDecorator = ['$delegate', function($delegate) {
+        });
+        // check if valid directives found
+        if (candidateDirectives.length === 0) {
+          throw new Error('No component found');
```
I think we should to use `$minErr` for these errors

drpicox committed:
feat(ngMock.$componentController): fix jscs extra space #13683

drpicox committed:
refactor(ngMock.$componentController): expect components to be restricted to 'E'  #13683

drpicox commented:
I tried to add minErr, 

```javascript
angular.mock.$ComponentControllerProvider = ['$compileProvider', function($compileProvider) {
  var minErr = angular.$$minErr('$componentController');
  return {
    $get: ['$controller','$injector', function($controller,$injector) {
      return function $componentController(componentName, locals, bindings, ident) {
        // get all directives associated to the component name
        var directives = $injector.get(componentName + 'Directive');
        // look for those directives that are components
        var candidateDirectives = directives.filter(function(directiveInfo) {
          // components have controller, controllerAs and restrict:'E'
          return directiveInfo.controller && directiveInfo.controllerAs && directiveInfo.restrict === 'E';
        });
        // check if valid directives found
        if (candidateDirectives.length === 0) {
          throw minErr('nocomp','No \'{0}\' component found', componentName);
        }
        if (candidateDirectives.length > 1) {
          throw minErr('multicomp','Found {0} directive candidates for component \'{1}\'', candidateDirectives.length, componentName);
        }
        // get the info of the component
        var directiveInfo = candidateDirectives[0];
        return $controller(directiveInfo.controller, locals, bindings, ident || directiveInfo.controllerAs);
      };
    }]
  };
}];
```

but I'm getting this error from `grunt test`:

```
error:   TypeError: Cannot read property 'nocomp' of undefined
    at /Users/drodenas/Projects/drpicox/angular.js/docs/config/processors/error-docs.js:43:76
    at Array.forEach (native)
    at Object.$process (/Users/drodenas/Projects/drpicox/angular.js/docs/config/processors/error-docs.js:18:12)
    at /Users/drodenas/Projects/drpicox/angular.js/node_modules/dgeni/lib/Dgeni.js:202:28
    at _fulfilled (/Users/drodenas/Projects/drpicox/angular.js/node_modules/dgeni/node_modules/q/q.js:798:54)
    at self.promiseDispatch.done (/Users/drodenas/Projects/drpicox/angular.js/node_modules/dgeni/node_modules/q/q.js:827:30)
    at Promise.promise.promiseDispatch (/Users/drodenas/Projects/drpicox/angular.js/node_modules/dgeni/node_modules/q/q.js:760:13)
    at /Users/drodenas/Projects/drpicox/angular.js/node_modules/dgeni/node_modules/q/q.js:821:14
    at flush (/Users/drodenas/Projects/drpicox/angular.js/node_modules/dgeni/node_modules/q/q.js:108:17)
    at doNTCallback0 (node.js:419:9)
```

I have added the directory `$componentController` with files `multicomp.ngdoc` and `nocomp.ngdoc` to `docs/content/error`.

I ran out of time. ¿Any clue how to solve it?

Thanks!

gkalpak commented:
Not sure what might be causing the issue. IF you push the changes (with `minErr`), even if broken, it will be easier to identify the problem (and the solution :smiley:).

drpicox committed:
refactor(ngMock.$componentController): add minErr to throw statements (fail due to TypeError: Cannot read property 'nocomp' of undefined)
    at /Users/drodenas/Projects/drpicox/angular.js/docs/config/processors/error-docs.js:43:76)

drpicox commented:
Thanks!
Done, breaking commit away.

petebacondarwin commented:
I'll take a look now

petebacondarwin commented:
The problem is that we were not minifying the `angular-mocks.js` file (unsurprisingly) and so the errors.json file did not contain the minErr error information, which is needed.

So looking at this again, I realise now that we don't use minErr in the rest of `ngMock` and so I was wrong to suggest that we use it here. Sorry.

petebacondarwin commented:
I will review and merge the rest of the PR without this last commit

drpicox commented:
No problem at all, I used throw new Error in the first approach because is what I found inside the angular-mock.js code.
Anyway it was nice to explore how minErr is working.

Thanks!
